stages/org.osbuild.gcp.guest-agent.conf

```diff
@@ -0,0 +1,259 @@
+      },
+      "startup": {
+        "type": "boolean",
+        "description": "Disables startup script execution."
+      },
+      "shutdown": {
+        "type": "boolean",
+        "description": "Disables shutdown script execution."
+      }
+    }
+  },
+  "NetworkInterfaces": {
+    "type": "object",
+    "additionalProperties": false,
+    "description": "NetworkInterfaces section.",
+    "minProperties": 1,
+    "properties": {
+      "setup": {
+        "type": "boolean",
+        "description": "Skips network interface setup."
+      },
+      "ip_forwarding": {
+        "type": "boolean",
+        "description": "Skips IP forwarding."
+      },
+      "dhcp_command": {
+        "type": "string",
+        "description": "Path for alternate dhcp executable used to enable network interfaces."
+      }
+    }
+  }
+},
+"additionalProperties": false,
+"required": ["config"],
+"description": "Configure GCP guest-agent.",
+"properties": {
+  "config_scope": {
+    "type": "string",
+    "description": "Create distro-wide or instance-specific configuration.",
+    "enum": ["distro", "instance"],
+    "default": "distro"
+  },
+  "config": {
+    "type": "object",
+    "additionalProperties": false,
+    "description": "GCP guest-agent configuration.",
+    "minProperties": 1,
+    "properties": {
+      "Accounts": {
+        "$ref": "#/definitions/Accounts"
+      },
+      "Daemons": {
+        "$ref": "#/definitions/Daemons"
+      },
+      "InstanceSetup": {
+        "$ref": "#/definitions/InstanceSetup"
+      },
+      "IpForwarding": {
+        "$ref": "#/definitions/IpForwarding"
+      },
+      "MetadataScripts": {
+        "$ref": "#/definitions/MetadataScripts"
+      },
+      "NetworkInterfaces": {
+        "$ref": "#/definitions/NetworkInterfaces"
+      }
+    }
+  }
+}
+"""
+
+def option_value_to_str(value):
+    """
+    Convert allowed types of option values to string.
+
+    The instance_configs.cfg allows comma seperated string lists as a option value.
+    """
+    if isinstance(value, list):
+        value = ", ".join(value)
+    elif isinstance(value, bool):
+        value = "true" if value else "false"
+    elif not isinstance(value, str):
+        value = str(value)
+    return value
+
+
+def main(tree, options):
+    scope = options.get("config_scope")
+    config = options.get("config")
+
+    instance_config_dir = f"{tree}/etc/default"
+    os.makedirs(instance_config_dir, exist_ok=True)
+
+    if scope == "distro":
+        filename = "instance_configs.cfg.distro"
+    else:
+        filename = "instance_configs.cfg"
+
+    parser = configparser.ConfigParser()
+
+    for section_id, section_content in config.items():
+        parser.add_section(section_id)
+        for key, value in section_content.items():
+            parser.set(section_id, key, option_value_to_str(value))
+
+    # ensure that we won't overwrite an existing file
+    with open(f"{instance_config_dir}/{filename}", "x") as f:
```

Comment:
Is there any particular reason why you chose to make the stage fail in case the file exists? Other stages tend to do two things AFAIK:

- Overwrite the file, regardless if it exists.
- Read the file content, modify it and write it back to the file. e.g.: https://github.com/osbuild/osbuild/blob/5fb7acc53e08b90bd039e74196c9e41d22566e02/stages/org.osbuild.dnf.config#L127-L147

if reading the file and modifying, we tend to use `iniparse`, instead of `configparser`, because it preserves the formatting of the original file, together with potential comments.

Reply:
I like the original approach. It is a safe approach without requiring reading the file back. Especially imagine the stage being used twice on the same tree (accidentally, or because pipelines are shared, etc.), this will catch the problem and error out, rather than silently overriding previous entries.

There is always an argument in favor of supporting overriding the previous content, or amending it. But as a safe fallback I like just erroring out.

Reply:
While I get your argument, I don't think that we should be preventing people from doing mistakes in manifests, by restricting what we can do with the stage.

By erroring, we may run into situation when Google (or even some distro) decides to ship a default configuration in their RPM and we will have to change the stage in order to be able to modify it. I think that failing if the file exists would make sense if we were creating drop-ins. But when touching a well-known config file path, I don't think that this is the best approach.

Reply:
Yeah, my thought process was "rather safe than sorry". In my mind there are a couple of scenarios:
- There is not existing configuration (distro or instance) -> a new config will be creaated
- A distro config exists and you try to create a new distro config -> throw an error and raise awareness that you are going to overwrite a preexisting configuration.
- A distro config exists (and user now knows about it) and creates an instance config which takes precedence over the distro config making it possible to overwrite the default config anyways.

But I do now see the complication that arise once someone ships a default configuration..

Reply:
Thanks for the explaining your though process Felix, this is helpful.

> Yeah, my thought process was "rather safe than sorry". In my mind there are a couple of scenarios:
> 
> * There is no existing configuration (distro or instance) -> a new config will be creaated
> * A distro config exists and you try to create a new distro config -> throw an error and raise awareness that you are going to overwrite a preexisting configuration.

The thing is that the user, nor someone maintaining the actual image definition in osbuild-composer, will have a way to solve this situation in case they want to override the file anyway. There is no way to delete arbitrary files from the filesystem tree as part of the image build.

Images built by osbuild are never booted, so there is no risk of overriding user's configuration.

> * A distro config exists (and user now knows about it) and creates an instance config which takes precedence over the distro config making it possible to overwrite the default config anyways.

Not sure how this would work in reality. osbuild and manifests are not intended for direct use by humans. In such case, the tool that assembles the manifest would have to first inspect a failed image build to determine that the issue is that a distro-wide configuration exists. If creating an instance-specific configuration would be a valid solution in such case (which it may not be for images intended for cloud provider's marketplace), the tool would have to generate a new, modified manifest with a different stage options. This would not scale, as the process would involve trying to build an image, let it fail and only then have enough information to assemble a working manifest to successfully build the image.

On a general note... While I do think that stages should prevent users from making errors, I think that there should be some limit. My expectation is that this should be done for example by not having too relaxed SCHEMA, etc.. I don't think that we should be really preventing users to use the stage in a wrong way - e.g. by combining it wrongly or have multiple instances of the stage touching the same config file. osbuild stages already make implicit assumptions, but do not restrict how users (or rather tools) can compose them into manifests. A good example of this is the RPM stage, and e.g. systemd stage enabling services. It is expected that you have to first install appropriate RPMs to be able to enable services.

I would like to summon @gicmo and hear his opinion as well.

Reply:
@FKolwa don't get me wrong, your reasons why you took this approach are definitely valid. I think that we just consider different use cases and situation when thinking about how the stage will be used. I may be wrong as well so it is important to discuss this properly and involve other people as well 😉